src/graph.h

```diff
@@ -13,26 +11,65 @@ namespace gs {
 class Graph : public torch::CustomClassHolder {
  public:
   Graph(bool is_subgraph) { is_subgraph_ = is_subgraph; }
+  Graph(bool is_subgraph, torch::optional<torch::Tensor> col_ids,
+        torch::optional<torch::Tensor> row_ids, int64_t num_cols,
+        int64_t num_rows)
+      : is_subgraph_{is_subgraph},
+        num_cols_{num_cols},
+        num_rows_{num_rows},
+        col_ids_{col_ids},
+        row_ids_{row_ids} {}
   void LoadCSC(torch::Tensor indptr, torch::Tensor indices);
+  void LoadCSR(torch::Tensor indptr, torch::Tensor indices);
   void LoadCSCWithColIds(torch::Tensor column_ids, torch::Tensor indptr,
                          torch::Tensor indices);
   void SetCSC(std::shared_ptr<CSC> csc);
+  void SetCSR(std::shared_ptr<CSR> csr);
+  void SetCOO(std::shared_ptr<COO> coo);
+  void SetData(torch::Tensor data);
+  void CSC2CSR();
+  void CSR2CSC();
   std::shared_ptr<CSC> GetCSC();
-  c10::intrusive_ptr<Graph> ColumnwiseSlicing(torch::Tensor column_ids);
-  c10::intrusive_ptr<Graph> RowwiseSlicing(torch::Tensor row_ids);
+  std::shared_ptr<CSR> GetCSR();
+  std::shared_ptr<COO> GetCOO();
+  torch::optional<torch::Tensor> GetData();
+  int64_t GetNumRows();
+  int64_t GetNumCols();
+  c10::intrusive_ptr<Graph> ColumnwiseSlicing(torch::Tensor column_index);
+  c10::intrusive_ptr<Graph> RowwiseSlicing(torch::Tensor row_index);
   c10::intrusive_ptr<Graph> ColumnwiseSampling(int64_t fanout, bool replace);
   c10::intrusive_ptr<Graph> ColumnwiseFusedSlicingAndSampling(
-      torch::Tensor column_ids, int64_t fanout, bool replace);
-  torch::Tensor RowIndices();
-  torch::Tensor AllIndices();
-  std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Relabel();
+      torch::Tensor column_index, int64_t fanout, bool replace);
+  torch::Tensor Sum(int64_t axis, int64_t powk);
+  torch::Tensor L2Norm(int64_t axis);
+  c10::intrusive_ptr<Graph> Divide(torch::Tensor divisor, int64_t axis);
+  c10::intrusive_ptr<Graph> Normalize(int64_t axis);
+  torch::Tensor AllValidNode();
+  torch::Tensor GetRows();
+  torch::Tensor GetCols();
+  torch::Tensor GetValidRows();
+  torch::Tensor GetValidCols();
+  torch::Tensor GetCOORows(bool is_original);
+  torch::Tensor GetCOOCols(bool is_original);
+  std::tuple<torch::Tensor, int64_t, int64_t, torch::Tensor, torch::Tensor,
+             torch::optional<torch::Tensor>, std::string>
+  Relabel();
   std::vector<torch::Tensor> MetaData();
 
-  void Print() const;
-
  private:
-  bool is_subgraph_;  // used for subgraph, stores node id in global graph
+  bool is_subgraph_;
+  int64_t num_cols_;  // total number of cols in a matrix
+  int64_t num_rows_;  // total number of rows in a matrix
   std::shared_ptr<CSC> csc_;
+  std::shared_ptr<CSR> csr_;
+  std::shared_ptr<COO> coo_;
+  torch::optional<torch::Tensor> data_;
+  torch::optional<torch::Tensor> col_ids_;  // column id in matrix
+  torch::optional<torch::Tensor> row_ids_;  // row id in matrix
```

Comment:
Do we store the global (original) eid in `Graph`? Or the eids in CSC/CSR are global eids?

Reply:
The eids in CSC/CSR are local eids. Currently we have not met the case where we need to fetch global e_id so we just haven't added global eid in `Graph`.

Reply:
I see